dss/stepfunctions/visitation/__init__.py

```diff
@@ -0,0 +1,295 @@
+                   for e in execs]
+
+        if len(running) < self.k_workers:
+            k_new = min(
+                self.k_workers - len(running),
+                len(self.waiting)
+            )
+
+            for i in range(k_new):
+                pfx = self.waiting[-1]
+                self._start_walker(pfx)
+                running.append(pfx)
+                del self.waiting[-1]
+
+        if running:
+            self.code = StatusCode.RUNNING.name
+        else:
+            self.code = StatusCode.SUCCEEDED.name
+
+        return running
+
+    def _start_walker(self, pfx: str) -> str:
+        """
+        Start a walker step function to work on a prefix.
+        """
+        name = walker_execution_name(self.name, pfx)
+        inp = {
+            'visitation_class_name': self.visitation_class_name,
+            'replica': self.replica,
+            'bucket': self.bucket,
+            'prefix': pfx,
+        }
+
+        try:
+            step_functions_invoke("dss-visitation-walker-{stage}", name, inp)
+        except botocore.exceptions.ClientError as ex:
+            raise DSSVisitationException() from ex
+
+        return name
+
+    def sentinel_initialize(self) -> None:
+        """
+        Impliment for initialization or sanity checking for a sentinel.
+        """
+        pass
+
+    def sentinel_finalize(self) -> None:
+        """
+        Impliment for finalization work for a succesful sentinel. Called once each worker has completed.
+        """
+        pass
+
+    def sentinel_finalize_failed(self) -> None:
+        """
+        Impliment for finalization work for a failed sentinel. This is your opportunity to cry, notify, and ruminate.
+        """
+        pass
+
+    def walker_initialize(self) -> None:
+        """
+        Impliment this method for initialization or sanity checking specifically for a walker.
+        """
+        pass
+
+    def process_item(self, key: str) -> None:
+        """
+        Subclasses must impliment this method. Called once for each blob visited.
+        """
+        raise NotImplementedError
+
+    def _walk(self) -> None:
+        """
+        Subclasses should not typically impliment this method, which includes logic specific to calling
+        self.process_item(*args) on each blob visited.
+        """
+
+        self.k_starts += 1
+
+        handle = Config.get_cloud_specific_handles(self.replica)[0]
+
+        blobs = handle.list_v2(
+            self.bucket,
+            prefix=self.prefix,
+            start_after_key=self.marker,  # type: ignore  # Cannot determine type of 'marker'
+            token=self.token  # type: ignore  # Cannot determine type of 'token'
+        )
+
+        self.code = StatusCode.RUNNING.name
+
+        for key in blobs:
+            self.process_item(key)
+            self.k_processed += 1
+            self.marker = blobs.start_after_key
+            self.token = blobs.token
+
```

Comment:
These two could be unified into a single opaque token. The fact that you distinguish between both is an implementation detail.

Reply:
Will think about it... this would require a change to cloud-blobstore.

Reply:
I'd suggest doing it in a separate PR, if you decide to do it.